File: leetcode/841. Keys and Rooms/main.go
// Solved by @kitanoyoru

package main

func canVisitAllRooms(rooms [][]int) bool {
	visited := make(map[int]bool, len(rooms))

	var dfs func(int)
	dfs = func(cur int) {
		visited[cur] = true
		for _, room := range rooms[cur] {
			if !visited[room] {
				dfs(room)
			}
		}
	}

	dfs(0)

	if len(visited) != len(rooms) {
		return false
	}

	return true
}

File: leetcode/841. Keys and Rooms/main.java
// Solved by @kitanoyoru
// https://leetcode.com/problems/keys-and-rooms/

class Solution {
  private void dfs(boolean[] visited, List<List<Integer>> rooms, int v) {
    visited[v] = true;
    for (int u : rooms.get(v)) {
      if (!visited[u]) {
        dfs(visited, rooms, u);
      }
    }
  }

  public boolean canVisitAllRooms(List<List<Integer>> rooms) {
    boolean[] visited = new boolean[rooms.size()];

    dfs(visited, rooms, 0);

    for (boolean val : visited) {
      if (!val) {
        return false;
      }
    }

    return true;
  }
}

